bootparameter sd card

hvilleneuve29 opened the pull request.

hvilleneuvedoo committed:
Add standalone boot parameters and BL2 binary files

bl2_bp.bin cannot be used in a SD card image. The bootrom expect the boot
parameters at sector 1, and the BL2 program at sector 8, so there is a gap of 6
sectors between the two. bl2_bp.bin has the boot parameters immediately followed
by BL2, so there is no such gap.

Add a separate bp.bin and bl2 files, so that they can be used in a WKS file to
generate a proper SD card image.

Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>

hvilleneuvedoo committed:
Add WKS file for rzg2l family SD card image

Note: this WKS file has no boot partition, as it is not needed. The boot
partition simply duplicates the same kernel image and DTB already found into the
rootfs partition inside the /boot directory.

Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>

patersonc commented on meta-rz-bsp/wic/rzg2lc-sdcard.wks:
```diff
@@ -0,0 +1,30 @@
+# short-description: Create SD card image for RZ/G2LC SoC
+# long-description: Creates a bootable SD card image including u-boot/atf + rootfs for Renesas RZ/G2LC SoC.
```
Hello @hvilleneuve29 

As far as I can tell, there is nothing here tying the wics file specifically for G2LC?  
Perhaps we can make this generic for the whole `g2l-family`?

Thanks, Chris

hvilleneuve29 replied:
Yes, I will modify the description, perhaps something like:
"Create SD card image for RZ/G2L family SoC"

hvilleneuve29 replied:
... and I will rename the file.

patersonc replied:
Thanks!

patersonc reviewed (commented):
Hello @hvilleneuve29 

Thank you for your pull request!

I'll kick off a build locally and test my side.

Regards, Chris

patersonc commented:
Hi @hvilleneuve29 

When I try to build with this wic file bitbake complains that `IMAGE_BOOT_FILES` is not set.  
Do you have the same issue?

My understanding is that files in `IMAGE_BOOT_FILES` will be put onto the `/boot` partition.  
What do you want to be added here?

Note:  
I built the wic image by adding the below to the machine conf:
```
IMAGE_FSTYPES:append = " wic.gz"
WKS_FILE = "rzg2lc-sdcard.wks"
```

Full error log;
```
ERROR: renesas-image-minimal-1.0-r0 do_image_wic: ExecutionError('/build/tmp/work/smarc_rzg2l-poky-linux/renesas-image-minimal/1.0-r0/temp/run.do_image_wic.113018', 1, None, None)
ERROR: Logfile of failure stored in: /build/tmp/work/smarc_rzg2l-poky-linux/renesas-image-minimal/1.0-r0/temp/log.do_image_wic.113018                          
Log data follows:                                                                                                                                              
| DEBUG: Executing python function extend_recipe_sysroot                                                                                                       
| NOTE: Direct dependencies are ['/build/../work/poky/meta/recipes-core/glibc/cross-localedef-native_2.35.bb:do_populate_sysroot', '/build/../work/poky/meta/recipes-core/glibc/ldconfig-native_2.12.1.bb:do_populate_sysroot', '/build/../work/poky/meta/recipes-devtools/binutils/binutils-cross_2.38.bb:do_populate_sysroot', '/build/../work/poky/meta/recipes-devtools/cdrtools/cdrtools-native_3.01.bb:do_populate_sysroot', '/build/../work/poky/meta/recipes-devtools/qemu/qemuwrapper-
cross_1.0.bb:do_populate_sysroot', '/build/../work/poky/meta/recipes-kernel/kmod/depmodwrapper-cross_1.0.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-core/update-rc.d/update-rc.d_0.8.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/btrfs-tools/btrfs-tools_5.16.2.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/createrepo-c/createrepo-c_0.19.0.bb:do_populate_sysroot', 'virtual:nat
ive:/build/../work/poky/meta/recipes-devtools/dnf/dnf_4.11.1.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/dosfstools/dosfstools_4.2.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/e2fsprogs/e2fsprogs_1.46.5.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/erofs-utils/erofs-utils_1.4.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/fdisk
/gptfdisk_1.0.8.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/makedevs/makedevs_1.0.1.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/mtools/mtools_4.0.38.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/opkg-utils/opkg-utils_0.5.0.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/opkg/opkg_0.5.0.bb:do_populate_sysroot', 'virtual:native:/
build/../work/poky/meta/recipes-devtools/pseudo/pseudo_git.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/rpm/rpm_4.17.1.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/squashfs-tools/squashfs-tools_git.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-devtools/syslinux/syslinux_6.04-pre2.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-extended/parted/part
ed_3.4.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-extended/pigz/pigz_2.7.bb:do_populate_sysroot', 'virtual:native:/build/../work/poky/meta/recipes-support/bmap-tools/bmap-tools_git.bb:do_populate_sysroot']
| NOTE: Installed into sysroot: []                                                                                                                             
| NOTE: Skipping as already exists in sysroot: ['cross-localedef-native', 'ldconfig-native', 'binutils-cross-aarch64', 'cdrtools-native', 'qemuwrapper-cross', 'depmodwrapper-cross', 'update-rc.d-native', 'btrfs-tools-native', 'createrepo-c-native', 'dnf-native', 'dosfstools-native', 'e2fsprogs-native', 'erofs-utils-native', 'gptfdisk-native', 'makedevs-native', 'mtools-native', 'opkg-utils-native', 'opkg-native', 'pseudo-native', 'rpm-native', 'squashfs-tools-native', 'sysli
nux-native', 'parted-native', 'pigz-native', 'bmap-tools-native', 'python3-native', 'sqlite3-native', 'file-native', 'glib-2.0-native', 'libxml2-native', 'curl-native', 'zlib-native', 'openssl-native', 'xz-native', 'expat-native', 'cmake-native', 'perl-native', 'util-linux-native', 'nasm-native', 'qemu-native', 'debianutils-native', 'shadow-native', 'kmod-native', 'texinfo-dummy-native', 'libtool-native', 'libsolv-native', 'libarchive-native', 'librepo-native', 'libcomps-nat
ive', 'gettext-minimal-native', 'libdnf-native', 'python3-iniparse-native', 'readline-native', 'ncurses-native', 'lzo-native', 'attr-native', 'lz4-native', 'zstd-native', 'lua-native', 'popt-native', 'bzip2-native', 'libgcrypt-native', 'elfutils-native', 'python3-setuptools-native', 'util-linux-libuuid-native', 'shared-mime-info-native', 'gnu-config-native', 'flex-native', 'python3-installer-native', 'python3-wheel-native', 'python3-six-native', 'coreutils-native', 'libnsl2-
native', 'libtirpc-native', 'gdbm-native', 'libffi-native', 'gettext-native', 'libpcre-native', 'make-native', 'libpcre2-native', 'libcap-ng-native', 'groff-native', 'gpgme-native', 'libcheck-native', 'libmodulemd-native', 'swig-native', 'json-c-native', 'gobject-introspection-native', 'libcap-native', 'libgpg-error-native', 'libmicrohttpd-native', 'itstool-native', 'm4-native', 'python3-flit-core-native', 'unzip-native', 'libassuan-native', 'libyaml-native', 'gnutls-native'
, 'nettle-native', 'gmp-native', 'libunistring-native', 'libidn2-native']                                                                                      
| DEBUG: Python function extend_recipe_sysroot finished                                                                                                        
| DEBUG: Executing python function set_image_size                                                                                                              
| DEBUG: 199981.600000 = 153832 * 1.300000                                                                                                                     
| DEBUG: 1150976.000000 = max(199981.600000, 1048576)[1048576.000000] + 102400                                                                                 
| DEBUG: 1150976.000000 = int(1150976.000000)                                                                                                                  
| DEBUG: 1150976 = aligned(1150976)                                                                                                                            
| DEBUG: returning 1150976                                                                                                                                     
| DEBUG: Python function set_image_size finished                                                                                                               
| DEBUG: Executing shell function do_image_wic                                                                                                                 
| INFO: Creating image(s)...                                                                                                                                   
|                                                                                                                                                              
| WARNING: bootloader config not specified, using defaults                                                                                                     
|                                                                                                                                                              
|                                                                                                                                                              
| ERROR: No boot files defined, IMAGE_BOOT_FILES unset for entry #4                                                                                            
| WARNING: exit code 1 from a shell command.                                                                                                                   
ERROR: Task (/build/../repo/meta-rz-bsp/recipes-core/images/renesas-image-minimal.bb:do_image_wic) failed with exit code '1'                                   
NOTE: Tasks Summary: Attempted 4009 tasks of which 4008 didn't need to be rerun and 1 failed.                                                                  
                                                                                                                                                               
Summary: 1 task failed:                                                                                                                                        
  /build/../repo/meta-rz-bsp/recipes-core/images/renesas-image-minimal.bb:do_image_wic                                                                         
Summary: There was 1 WARNING message.                                                                                                                          
Summary: There was 1 ERROR message, returning a non-zero exit code. 
```

hvilleneuve29 commented:
Hi Chris,
I set the required variables in another layer like this:

IMAGE_FSTYPES += "wic.gz wic.bmap"

WKS_FILE = "rzg2lc-sdcard.wks"

IMAGE_INSTALL:append = " \
    kernel-devicetree \
    kernel-modules \
    kernel-image \
"

IMAGE_BOOT_FILES = " \
    Image \
    rzg2lc-hummingboard.dtb \
"

patersonc commented:
Hello @hvilleneuve29 

I've created a patch that adds the WKS_FILE and other related settings. I've also renamed the wfs file that you've created as we discussed and added your SOB to your patches.

Do you mind checking the "Allow edits from maintainers" box so I can push these changes directly to your branch?

Then we can get this PR merged :)

Thanks, Chris

hvilleneuve29 commented:
Hi Chris,
I have already modified, like I said, the WKS file. I have also removed the /boot partition, since it is not needed when using EXTLINUX. Let me test my changes and I will update my branch with all of the changes, and after that you can add your patches?

hvilleneuvedoo committed:
kernel: simplify device tree declaration

Also to be able to reference the devicetree basename in extlinux for example
(future commit).

Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>

hvilleneuvedoo committed:
Add extlinux configuration support

To be able to use distroboot. If U-Boot is not configured for distroboot,
adding it to the image will be harmless.

Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>

hvilleneuvedoo committed:
rzg2l: enable SD card WKS file for g2l family SOCs

Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>

hvilleneuve29 commented:
Hi Chris,
I have updated the bootparameter_sd_card branch with the updated WKS file, as well as some other commits to fix SD card boot. I have also selected the default as SD card boot, maybe you will want to change this.

On top of this branch, I have created a new branch extlinux, which is needed to boot the solidrun EVK as it uses distroboot/extlinux.

And finally there is the branch solidrun, which adds support for the solidrun EVK.

I have tested all these branches on both the solidrun EVK and renesas rzg2lc EVK (using the renesas-image-minimal image).

Let me know what you think.
Hugo.

patersonc commented:
Hello @hvilleneuve29 

Thank you for the updates.

> Hi Chris, I have updated the bootparameter_sd_card branch with the updated WKS file, as well as some other commits to fix SD card boot. I have also selected the default as SD card boot, maybe you will want to change this.

I think it's okay to set the SD card to be the default boot device. Although I'm guessing that we'd need to make further changes if we wanted to save the u-boot environment to the SD card?

If users want to use tftp/nfs then they'll likely be flashing the bootloaders to the QSPI anyway, and they can update the u-boot env accordingly.

patersonc commented on meta-rz-bsp/recipes-bsp/u-boot/files/0002-configs-smarc-rzg2l-g2lc-fix-SD-card-boot-and-set-it.patch:
```diff
@@ -0,0 +1,91 @@
+-	"emmcload=ext4load mmc 0:2 0x48080000 boot/Image;ext4load mmc 0:2 0x48000000 boot/r9a07g044l2-smarc.dtb;run prodemmcbootargs \0" \
+-	"sd1load=ext4load mmc 1:2 0x48080000 boot/Image;ext4load mmc 1:2 0x48000000 boot/r9a07g044l2-smarc.dtb;run prodsdbootargs \0" \
+-	"bootcmd_check=if mmc dev 1; then run sd1load; else run emmcload; fi \0"
++	"mmcload=ext4load mmc ${mmc_dev}:${root_part} 0x48080000 boot/Image;ext4load mmc ${mmc_dev}:${root_part} 0x48000000 boot/r9a07g044c2-smarc.dtb; run prodbootargs\0"
```
This should be `r9a07g044l2-smarc.dtb` for G2L.

hvilleneuve29 replied:
Ok, will fix it.

patersonc commented on meta-rz-bsp/conf/machine/include/rz-common.inc:
```diff
@@ -1,7 +1,7 @@
 
 MACHINE_FEATURES = ""
-IMAGE_FSTYPES:append = " tar.gz"
+IMAGE_FSTYPES:append = " tar.gz wic.gz wic.bmap"
```
We should move this change to `meta-rz-bsp/conf/machine/include/rzg2l-family.inc` until we add support for the `hihope-rzg2h` machine, otherwise we'll get a build error complaining that `WKS_FILE` isn't set.

hvilleneuve29 replied:
You are right, will do.

patersonc commented:
Please could you add your SOB (signed-off-by) to your commits, and to the patches you've created for u-boot?

patersonc commented:
Maybe we could enable u-boot env saving to the SD card?  
https://renesas.info/wiki/RZ-G/RZ-G2L_SMARC#Use_SD_card_for_u-boot_environment_variables

patersonc commented:
It would also be good to add the same u-boot env changes for the smarc-rzg2ul platform.

hvilleneuve29 commented:
I will have a look at u-boot env saving to the SD card...
Thank you for the review.

hvilleneuvedoo committed:
u-boot: rename patches folder to better differentiate origin

The patch is specific to u-boot, not u-boot-renesas or other.

Move it into an appropriately named folder, to better differentiate where each
patch belongs.

Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>

hvilleneuvedoo committed:
rzg2: add U-Boot distroboot support

Uses extlinux to simplify boot process.

Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>

hvilleneuvedoo committed:
rzg2: add support for environment in SD card

Allow to save environment to SD card.

Signed-off-by: Hugo Villeneuve <hvilleneuve@dimonoff.com>

hvilleneuve29 commented:
Hi Chris,
I updated the bootparameter_sd_card branch to incorporate all the changes that we discussed. I decided to use distroboot/extlinux, as this is now the recommended way to setup U-Boot booting process. This is also the default method for SolidRun EVK.

I added support to save the environment to SD card. Tested with smarc-rzg2lc EVK board. I did not set the bus width to 4 as stated in the wiki, as it will revert to 1 bit automatically in this case, from what I understand, and speed is not really an issue reading/writing such a small block of data (setting it to 4 would reduce emmc performance).

I tested the compilation of all other machines in the layer with success.

I added my SOB tags to all commits, including those in U-Boot patches.

Hugo.

hvilleneuve29 commented:
On Tue, 23 Apr 2024 02:44:06 -0700
Chris Paterson ***@***.***> wrote:

> @patersonc commented on this pull request.
> 
> 
> 
> > + arch/arm/dts/smarc-rzg2l.dts   | 2 +-
> + arch/arm/dts/smarc-rzg2lc.dts  | 2 +-
> + arch/arm/dts/smarc-rzg2ul.dts  | 2 +-
> 
> As far as I can see, we're not actually touching these files? Could you regenerate the patch?

Hi,
yes, sorry for that, will do.

 
> Could you also add your SOB?

Yes.

Hugo.
